### PR TITLE
validate number and cvc for credit cards

### LINKED ISSUE #36: Update Credit Card Validations
Currently:

```ruby
2.3.0 > user.gateway.credit_cards.create!
ActiveRecord::RecordInvalid: Validation failed:
Name can't be blank,
Last digits can't be blank,
Month can't be blank,
Year can't be blank
```

These error messages aren't that useful. It might be better if it read:

```ruby
2.3.0 > user.gateway.credit_cards.create!
ActiveRecord::RecordInvalid: Validation failed:
Name can't be blank,
Number can't be blank,
Month can't be blank,
Year can't be blank,
Cvc can't be blank
```

This way a user knows they have to provide `:number` and `:cvc`, instead of just `:last_digits`.

### PR DESCRIPTION
**Pull request by BenMorganIO**
Before:

```ruby
2.3.0 :007 > user.gateway.credit_cards.create!
   (0.1ms)  begin transaction
   (0.1ms)  rollback transaction
ActiveRecord::RecordInvalid: Validation failed: Name can't be blank, Last digits can't be blank, Month can't be blank, Year can't be blank
2.3.0 :008 > user.gateway.credit_cards.create!(name: 'Ben Morgan', number: 4111_1111_1111_1111, month: 8, year: 2019)
   (0.1ms)  begin transaction
  SQL (1.8ms)  INSERT INTO "spur_credit_cards" ("name", "last_digits", "brand", "month", "year", "token", "gateway_id", "created_at", "updated_at") VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)  [["name", "Ben Morgan"], ["last_digits", "1111"], ["brand", "Visa"], ["month", 8], ["year", 2019], ["token", "card_17tlYu2BZSQJcNSQzETZitDO"], ["gateway_id", 2], ["created_at", 2016-03-27 21:36:44 UTC], ["updated_at", 2016-03-27 21:36:44 UTC]]
   (3.6ms)  commit transaction
 => #<Spur::PaymentMethod::CreditCard id: 2, name: "Ben Morgan", last_digits: "1111", brand: "Visa", month: 8, year: 2019, token: "card_17tlYu2BZSQJcNSQzETZitDO", gateway_id: 2, created_at: "2016-03-27 21:36:44", updated_at: "2016-03-27 21:36:44">
```

After:

```ruby
2.3.0 :007 > user.gateway.credit_cards.create!
   (0.1ms)  begin transaction
   (0.1ms)  rollback transaction
ActiveRecord::RecordInvalid: Validation failed: Name can't be blank, Number can't be blank, Month can't be blank, Year can't be blank, Cvc can't be blank
2.3.0 :008 > user.gateway.credit_cards.create!(name: 'Ben Morgan', number: 4111_1111_1111_1111, month: 8, year: 2019)
   (0.1ms)  begin transaction
   (0.1ms)  rollback transaction
ActiveRecord::RecordInvalid: Validation failed: Cvc can't be blank
2.3.0 :009 > user.gateway.credit_cards.create!(name: 'Ben Morgan', number: 4111_1111_1111_1111, month: 8, year: 2019, cvc: 123)
   (0.1ms)  begin transaction
  SQL (1.8ms)  INSERT INTO "spur_credit_cards" ("name", "last_digits", "brand", "month", "year", "token", "gateway_id", "created_at", "updated_at") VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)  [["name", "Ben Morgan"], ["last_digits", "1111"], ["brand", "Visa"], ["month", 8], ["year", 2019], ["token", "card_17tlYu2BZSQJcNSQzETZitDO"], ["gateway_id", 2], ["created_at", 2016-03-27 21:36:44 UTC], ["updated_at", 2016-03-27 21:36:44 UTC]]
   (3.6ms)  commit transaction
 => #<Spur::PaymentMethod::CreditCard id: 2, name: "Ben Morgan", last_digits: "1111", brand: "Visa", month: 8, year: 2019, token: "card_17tlYu2BZSQJcNSQzETZitDO", gateway_id: 2, created_at: "2016-03-27 21:36:44", updated_at: "2016-03-27 21:36:44">
```

---

fixes #36

### CODE REVIEW

File: ryal/app/models/spur/payment_method/credit_card.rb [hunk 1]
```diff
@@ -3,9 +3,9 @@ module PaymentMethod
     class CreditCard < Spur::PaymentMethod::Base
       attr_accessor :number, :cvc
 
-      validates :name, :last_digits, :month, :year, :gateway_id, presence: true
+      validates :name, :number, :month, :year, :gateway_id, :cvc, presence: true
 
-      before_validation do
+      after_validation do
```
Comment (hickscorp) [inline ryal/app/models/spur/payment_method/credit_card.rb]: I'm kinda :-1: on this. What's the reason we're doing this in `after_validation`?

Comment (BenMorganIO) [inline ryal/app/models/spur/payment_method/credit_card.rb]: I feel like Rails should validate if a `number` is present, not if there's `last_digits`. When we update and create a credit card, we want to supply the whole number, not just the 4 last digits...

Its a bit confusing..

Comment (BenMorganIO) [inline ryal/app/models/spur/payment_method/credit_card.rb]: The `after_validation` assumes that there's a number field. To make sure its present, we have to run the validations...

Comment (hickscorp) [inline ryal/app/models/spur/payment_method/credit_card.rb]: Note for self: instead do:
```
def number=(val)
  super val
  self[:last_digit] = val.to_s.gsub(/[[:space:]]/, '').chars.last(4).join
  val
end

def number=(*)
  Rails.logger.warn 'Using #number= is not allowed, and has no effect.'
end
```

Comment (BenMorganIO) [inline ryal/app/models/spur/payment_method/credit_card.rb]: What about:

```ruby
def number=(value)
  self[:last_digit] = value.to_s.gsub(/[[:space:]]/, '').chars.last(4).join
  super value
end

def last_digits=(*)
  Rails.logger.warn 'Using #last_digits= is not allowed, and has no effect.'
end
```

Comment (BenMorganIO) [inline ryal/app/models/spur/payment_method/credit_card.rb]: Does `number=` get called in an update?

Comment (hickscorp) [inline ryal/app/models/spur/payment_method/credit_card.rb]: Initially wrote that, but then realized that super has to come first in the method. Maybe it's not the case in ruby tho :D

Comment (hickscorp) [inline ryal/app/models/spur/payment_method/credit_card.rb]:
```ruby
attr_accessor :number

def number=(value)
  write_attribute :last_digit, value.to_s.gsub(/[[:space:]]/, '').chars.last(4).join
  @number = value
end
```

File: ryal/spec/factories.rb [hunk 2]
```diff
@@ -14,6 +14,7 @@
     number { Array.new(16) { rand 10 }.join.to_i }
     month Time.current.month
     year Time.current.year + 1
```
Comment (hickscorp) [inline ryal/spec/factories.rb]: Hound's disabled on this project, isn't it? :D

Comment (BenMorganIO) [inline ryal/spec/factories.rb]: Yep

File: ryal/app/models/spur/payment_method/credit_card.rb [hunk 3]
```diff
@@ -1,15 +1,23 @@
...
-      validates :name, :last_digits, :month, :year, :gateway_id, presence: true
+      validates :name, :number, :month, :year, :gateway_id, :cvc, presence: true
 
-      before_validation do
-        self.last_digits = number.to_s.gsub(/[[:space:]]/, '').chars.last(4).join
+      before_create -> { gateway.create_credit_card self }
+
+      def number=(value)
+        self[:last_digits] = value.to_s.gsub(/[[:space:]]/, '').chars.last(4).join
+        @number = value
```
Comment (BenMorganIO) [inline ryal/app/models/spur/payment_method/credit_card.rb]: > The card number, as a string without any separators.

- https://stripe.com/docs/api?lang=ruby#create_card-source-number

I'm thinking we should do:

```ruby
@number = value.to_s.gsub(/[[:space:]]/, '')
```

**Pull request merged**